private/buf/cmd/buf/command/beta/studioagent/studioagent.go

```diff
@@ -0,0 +1,165 @@
+
+package studioagent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/bufpkg/bufstudioagent"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/interrupt"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"go.uber.org/zap"
+)
+
+const (
+	portFlagName              = "port"
+	disallowedHeadersFlagName = "disallowed-header"
+	forwardHeadersFlagName    = "forward-header"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name + " <origin>",
+		Short: "Run an HTTP server as the studio agent with the origin be set as the allowed origin for CORS options.",
+		Args:  cobra.ExactArgs(1),
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	Port              string
+	DisallowedHeaders []string
+	ForwardHeaders    []string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.Port,
+		portFlagName,
+		"8080",
+		"The port to be exposed to accept HTTP request.",
+	)
+	flagSet.StringSliceVar(
+		&f.DisallowedHeaders,
+		disallowedHeadersFlagName,
+		nil,
+		`The headers to be disallowed via the agent to the target server. Multiple headers are appended if specified multiple times.`,
+	)
+	flagSet.StringSliceVar(
+		&f.ForwardHeaders,
+		forwardHeadersFlagName,
+		nil,
+		`The headers to be forwarded via the agent to the target server. Must be a colon-separated key-value pair (like --forward-header=fromHeader1:toHeader1). Multiple header pairs are appended if specified multiple times.`,
+	)
+}
+
+func run(
+	ctx context.Context,
+	container appflag.Container,
+	flags *flags,
+) error {
+	logger := container.Logger().Named("studio-agent")
+	// convert the disallowedHeaders from a list of string to a map
+	disallowedHeaders := make(map[string]struct{}, len(flags.DisallowedHeaders))
+	for _, header := range flags.DisallowedHeaders {
+		disallowedHeaders[header] = struct{}{}
+	}
+	// convert the forwardHeaders from a list of colon-separated key-value pair string to a map
+	forwardHeaders := make(map[string]string, len(flags.ForwardHeaders))
+	for _, pair := range flags.ForwardHeaders {
+		s := strings.Split(pair, ":")
+		if len(s) != 2 {
+			return fmt.Errorf("unknown key-pair value of forward-headers: %s", pair)
+		}
+		forwardHeaders[s[0]] = s[1]
+	}
+	config, err := bufcli.NewConfig(container)
+	if err != nil {
+		return err
+	}
+	mux := bufstudioagent.NewHandler(
+		logger,
+		container.Arg(0), // the origin from command argument
+		config.TLS,
```

Comment:
Where does this TLS come from, is this some global config option you can set over the whole binary?

I imagine this was something to configure the TLS for outgoing calls made by the cli towards a bsr, which we could re-use here if we document it well. Besides the outgoing tls config, I think we also need to let users set a server TLS config to be used by the `listenAndServe{TLS}` used in this subcommand.

Reply:
added flags for ca cert, client cert&key and server cert&key file, instead of the global TLS config